File: 3068/solution.py
class Solution:
    def maximumValueSum(self, nums: list[int], k: int, edges: list[list[int]]) -> int:
        res = sum(nums)
        num_nodes = 0
        smallest_diff = float("inf")
        smallest_drop = float("inf")
        for num in nums:
            if num ^ k > num:
                diff = (num ^ k) - num
                res += diff
                smallest_diff = min(smallest_diff, diff)
                num_nodes += 1
            else:
                smallest_drop = min(smallest_drop, num - (num ^ k))
        if num_nodes % 2 == 0:
            return res
        return res - min(smallest_diff, smallest_drop)
        # time O(n)
        # space O(1)

File: 3068/solution.java
class Solution {
    public long maximumValueSum(int[] nums, int k, int[][] edges) {
        int totalNums = 0;
        long res = 0;
        int smallestDecrease = Integer.MAX_VALUE;
        int smallestIncrease = Integer.MAX_VALUE;

        for (int num : nums) {
            if ((num ^ k) > num) {
                totalNums++;
                res += num ^ k;
                smallestIncrease = Math.min(smallestIncrease, (num ^ k) - num);
            } else {
                res += num;
                smallestDecrease = Math.min(smallestDecrease, num - (num ^ k));
            }
        }

        if (totalNums % 2 == 0) {
            return res;
        } else {
            return Math.max(res - smallestIncrease, res - smallestDecrease);
        }
    }
}

// time O(n)
// space O(1)